### ISSUE TITLE
IdTCPServer dead connection, cannot remove from Contexts list, then hangs 100% CPU on shutdown

### ISSUE THREAD

**Issue by SARTrack** [IndySockets/Indy#484]
This is an intermittent but disastrous problem, which under Linux occured all the time, until I gave up on this. Now this problem is occuring on a Windows Server VPS, and when it happend I could no longer get into the VPS due to the TCPServer looping somewhere at 100% CPU after shutdown.
I have now been able to reproduce the issue a few times on a test server, by connecting to the server via the StarLink system, while previousely I could (in Linux) only reproduce it on a production server on the other side of the planet.

What seems to happen is this:
A TCP connection to the TCPServer somehow goes into limbo. That is, there is no TCP RESET, somehow the connection gets severed without a RESET... I think this issue occurs a lot more using the StarLink system, probably during a satellite hand-over which goes wrong.
BUT something *is* happening on the server side, because somehow the connection dissapears, but the Context is not freed, and becomes a dead Context  in the List, and it cannot be removed.
In my case this means that the specific Client can no longer connect, because the system thinks the Client is still connected.
Then, after a Timeout event, the system attemps to Disconnect the Client, but this fails to remove it from the list.
Below I show a Debug Log of what happens.

Client ID is "TESTPH". I call TIdContext(List[Index]).Connection.Disconnect in the LockList:

[00:04:47/059Z] IdTCPConnection.Disconnect calls DisConnect(True)
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer)
[00:04:47/059Z] IdIOHandlerSocket.Close
[00:04:47/059Z] IdSocketHandle.CloseSocket: HandleAllocated
[00:04:47/059Z] IdSocketHandle.Disconnect
[00:04:47/059Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[00:04:47/059Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
[00:04:47/059Z] IdIOHandler.Close LIntercept is NIL
[00:04:47/059Z] IdComponent.DoStatus
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): NOT Connected
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is assigned
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is NOT Opened
[00:04:47/059Z] DisconnectClient: Port=2: AFTER List: 0: TESTPH Timestamp=0
[00:04:47/160Z] Superuser: KeepAliveTimerEvent (Port2): List.Count=1
[00:04:47/160Z] ERROR: KeepAliveTimerEvent (Port2): DTBClient was not released! Index=0
[00:04:47/160Z] 0  Final List: TESTPH

What *should* have happened (Another Client ID "BARTPH"):

D3: [00:04:17/019Z] DisconnectClient: Disconnecting: Port=2  Callsign=BARTPH Index=1  Count=2
D3: [00:04:17/019Z] DisconnectClient: Port=2: Before List: 0: TESTPH Timestamp=0
D3: [00:04:17/019Z] DisconnectClient: Port=2: Before List: 1: BARTPH Timestamp=0
[00:04:17/019Z] IdTCPConnection.Disconnect calls DisConnect(True)
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer)
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer) calls DisconnectNotifyPeer
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is assigned
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is Opened: Calling DoStatus(Disconnecting)
[00:04:17/019Z] IdComponent.DoStatus
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer) calls LIOHandler.Close
[00:04:17/019Z] IdIOHandlerSocket.Close
[00:04:17/019Z] IdSocketHandle.CloseSocket: HandleAllocated
[00:04:17/019Z] IdSocketHandle.Disconnect
[00:04:17/019Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[00:04:17/019Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
[00:04:17/019Z] IdIOHandler.Close LIntercept is NIL
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer) calls DoOnDisconnected
[00:04:17/019Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is Opened: Calling DoStatus(hsDisconnected)
[00:04:17/019Z] IdComponent.DoStatus
[00:04:17/049Z] IdTCPServer2Disconnect: Client Disconnected: BARTPH
[00:04:17/067Z] IdContext.Destroy calls FContextList.Remove(Self)
[00:04:17/068Z] IdContext.Destroy calls IdDisposeAndNil(FConnection)
[00:04:17/068Z] IdIOHandlerSocket.Close
[00:04:17/069Z] IdSocketHandle.CloseSocket: NOT HandleAllocated
[00:04:17/069Z] IdIOHandler.Close: Host=  Connected=0
[00:04:17/069Z] IdIOHandler.Close LIntercept is NIL
[00:04:17/070Z] IdSocketHandle.Destroy: CloseSocket, Free FConnectionHandle, Free FReadSocketList
[00:04:17/070Z] IdSocketHandle.CloseSocket: NOT HandleAllocated
[00:04:17/071Z] IdSocketList.Destroy
[00:04:17/071Z] IdIOHandler.Close: Host=  Connected=0
[00:04:17/071Z] IdIOHandler.Close LIntercept is NIL

So what I see here, in the first example:
IdTCPConnection.Disconnect(ANotifyPeer): NOT Connected

Somehow the TCP connection got disconnected, but not in a normal way.
In the current Debug log I see no event which shows the TCP connection being dropped.

I do not know where I should look, to be able to monitor an TCP event, that is, when a Connection is dropped from the outside, where do I look to trace the steps backward, and to see why it is not properly also removing the IdContext from the list?
And, where does the TCPServer causes this 100% CPU loop which takes the entire VPS down?
This happens at: IdTCPServer2.Active := false;

I don't understand much of the code. I am not that good a programmer, but I am happy to produce more Debug logs, if somebody can tell me where to look.

NOTE: This issue will very likely also affect higher level protocols, like the HTTPServer which also causes weird hang-ups when KeepAlive is TRUE.

**Comment by rlebeau** [comment]
> when it happend I could no longer get into the VPS due to the TCPServer looping somewhere at 100% CPU after shutdown.

That usually means there is a runaway thread running a tight unyielding loop, thus eating up CPU cycles.

> What seems to happen is this...

Describing the code's behavior is not helpful. Please show your actual code.

> I call TIdContext(List[Index]).Connection.Disconnect in the LockList:

Are you unlocking the `Contexts` list? Also, when closing sockets across thread boundaries, consider using `TIdContext.Binding.CloseSocket()` instead of `TIdContext.Connection.Disconnect()`.  Let the owning thread handle the actual closure of the socket.

> Somehow the TCP connection got disconnected, but not in a normal way. In the current Debug log I see no event which shows the TCP connection being dropped.

There is no event.  Such Indy uses blocking sockets, a *graceful* disconnect is detected when a read operation reports 0 bytes received, and an *abnormal* disconnect is detected when a read operation reports an error.  The server's `OnDisconnect` event is fired when the socket's owning thread is terminating.

> And, where does the TCPServer causes this 100% CPU loop which takes the entire VPS down? This happens at: IdTCPServer2.Active := false;

Deactivating the server will close all open sockets, and then wait in a loop for their owning threads to terminate and clear their entries from the `Contexts` list.  That wait loop sleeps in between each loop iteration, so it does not eat up CPU cycles. However, a runaway socket thread that is not responding to terminate requests certainly can.

**Comment by SARTrack** [comment]
I did not see your reply until I posted on atozed.com just now;
The Code is the the TCPServer code.  You should have all that. Do you want me to post all the many bits where I now have a Log entry?
Yes, I properly lock and unlock the list. And I do it again to see the result (after a short delay) so I can see the entry was not removed.

The problem occurs before. When I discover a Timeout on the connection, this means that by that time the connection has already been severed. But the Server has not properly removed the Context.
Then I try to force a Disconnect, and in the Log you can see that this fails because:

[00:04:47/059Z] IdComponent.DoStatus
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): NOT Connected <<<<<<<
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is assigned
[00:04:47/059Z] IdTCPConnection.Disconnect(ANotifyPeer): LIOHandler is NOT Opened <<<<<<

So nothing happens, and the Context is not freed.

I will change the code to TIdContext.Binding.CloseSocket() but I think this will have no effect.
I will report back.

**Comment by SARTrack** [comment]
If a TCP event occurs (a TCP RESET for example), *where* in the code can I add a debug log entry to report this?

**Comment by SARTrack** [comment]
Okay, this time it happened almost immediately:

D3: [22:37:02/087Z] [12345] KeepAliveTimerEvent: Port 2: (NOT AllSyncsComplete): Timeout for Client: Callsign=TESTPH, IP=192.168.50.110 PongCounter=7 Disconnecting on Index: 0
D3: [22:37:02/087Z] DisconnectClient: Disconnecting: Port=2  Callsign=TESTPH Index=0  Count=1
D3: [22:37:02/087Z] DisconnectClient: Port=2: Before List: 0: TESTPH Timestamp=0
[22:37:02/087Z] IdSocketHandle.Disconnect
[22:37:02/087Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[22:37:02/087Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
D3: [22:37:02/287Z] DisconnectClient: Port=2: AFTER List: 0: TESTPH Timestamp=0
D3: [22:37:02/387Z] Superuser: KeepAliveTimerEvent (Port2): List.Count=1
ERROR: [22:37:02/387Z] [12345] KeepAliveTimerEvent (Port2): DTBClient was not released! Index=0
D3: [22:37:02/387Z] Superuser: 0  Final List: TESTPH
[22:37:07/096Z] IdSocketHandle.CloseSocket: NOT HandleAllocated
D3: [22:37:07/096Z] [12345] KeepAliveTimerEvent: Port 2: (NOT AllSyncsComplete): Timeout for Client: Callsign=TESTPH, IP=192.168.50.110 PongCounter=8 Disconnecting on Index: 0
D3: [22:37:07/096Z] DisconnectClient: Disconnecting: Port=2  Callsign=TESTPH Index=0  Count=1
D3: [22:37:07/096Z] DisconnectClient: Port=2: Before List: 0: TESTPH Timestamp=0
D3: [22:37:07/298Z] DisconnectClient: Port=2: AFTER List: 0: TESTPH Timestamp=0
D3: [22:37:07/400Z] Superuser: KeepAliveTimerEvent (Port2): List.Count=1
ERROR: [22:37:07/400Z] [12345] KeepAliveTimerEvent (Port2): DTBClient was not released! Index=0
D3: [22:37:07/400Z] Superuser: 0  Final List: TESTPH

I can show you the code, what is the best place to discuss this, in Atozed or here? In Atozed I can attach files.

**Comment by SARTrack** [comment]
I cleaned up a bit for clarity. This is the last test:

LAST TIME IT WAS ALIVE:
[23:36:36/074Z] [12345] *** ProcessIncoming: Calling DTBTransmitEnableTimeout for App: TESTPH

(I do not see anything related to this connection in the meantime. No exceptions, anything.)

Then:
FIRST ATTEMPT TO DISCONNECT after 30 seconds no PINGs received (dead client):

[23:37:08/948Z] [12345] KeepAliveTimerEvent: Port 2: Timeout for Client: Callsign=TESTPH, IP=192.168.50.110 PongCounter=7 Disconnecting on Index: 0
[23:37:08/948Z] DisconnectClient: Disconnecting: Port=2  Callsign=TESTPH Index=0  Count=1
[23:37:08/948Z] DisconnectClient: Port=2: Before List: 0: TESTPH Timestamp=0
[23:37:08/948Z] DisconnectClient calls TIdContext(List[Index]).Binding.CloseSocket  <<<<<<<
[23:37:08/948Z] IdSocketHandle.Disconnect
[23:37:08/948Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[23:37:08/948Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
[23:37:13/947Z] IdSocketHandle.CloseSocket: NOT HandleAllocated, Doing nothing. <<<<<<<<<<<<<<<
[23:37:09/048Z] Superuser: KeepAliveTimerEvent (Port2): List.Count=1
ERROR: [23:37:09/048Z] [12345] KeepAliveTimerEvent (Port2): DTBClient was not released! Index=0
[23:37:09/048Z] Superuser: 0  Final List: TESTPH

I need to know where to look for the event which 'disconnected' the socket. And to see what socket error (if any) was generated.
It must be something special, because a 'normal' socket disconnect should (and is) always properly handled bt TCPServer.

**Comment by rlebeau** [comment]
> The Code is the the TCPServer code. You should have all that.

Are you referring to a demo or something?

> Do you want me to post all the many bits where I now have a Log entry?

Yes, I need to see your actual server code (or at least a minimal demo) that is producing your log.  Your log is very difficult to follow without seeing the server logic behind it.

> The problem occurs before. When I discover a Timeout on the connection, this means that by that time the connection has already been severed. But the Server has not properly removed the Context.

Typically, this happens when a server event handler is not coded properly and is preventing the calling socket thread from terminating properly.  For instance, the most common offender is catching Indy exceptions and discarding them instead of letting `TIdTCPServer` handle them. Since the socket thread is looped, a failure to let the thread terminate properly can lead to a runaway thread, thus the higher CPU usage.  The `OnDisconnect` event is not fired, and the `TIdContext` object is not removed from the `Contexts` list, until the socket thread is shutting down.

> Then I try to force a Disconnect and in the Log you can see that this fails because:
> ...
> I will change the code to TIdContext.Binding.CloseSocket() but I think this will have no effect.

When disconnecting a client across thread boundaries, it is safer to use `Binding.CloseSocket()` instead of `TIdTCPConnection.Disconnect()`.  The former will just close the underlying socket, whereas the latter will close (and potentially free) the `IOHandler`, which has multi-threading risks.

I start to question why you are disconnecting clients from outside of the server threads in the first place (ie, why are you running an external timer)?  Are you not able to just use timeouts directly in the server's events instead?

> If a TCP event occurs (a TCP RESET for example), _where_ in the code can I add a debug log entry to report this?

It won't be reported until the *next* I/O operation on the socket sees the event and reports it back to Indy as a socket error.

> I can show you the code, what is the best place to discuss this, in Atozed or here? In Atozed I can attach files.

You should be able to attach files here.

**Comment by SARTrack** [comment]
> > The Code is the the TCPServer code. You should have all that.
> 
> Are you referring to a demo or something?

No, I am refering to the actual source code of the IdTCPServer and all the other Indy10 units it uses. And to trace down what calls what is for me nearly impossible in Indy10. Especially when tracing through the code runs into a dead-end of 'Virtual'. After which I have to do manual search through all the Indy code to find any places where that 'virtual' code is being used, then hope I have the right one to put my debug Log bits in.

> > The problem occurs before. When I discover a Timeout on the connection, this means that by that time the connection has already been severed. But the Server has not properly removed the Context.
 
> Typically, this happens when a server event handler is not coded properly and is preventing the calling socket thread from terminating properly. For instance, the most common offender is catching Indy exceptions and discarding them instead of letting `TIdTCPServer` handle them.

I 'debug Log' all exceptions I get from the TCPServer. In this case I do not see any. Also, are you telling me that I should 'raise' any and all Indy10 exeptions after my program gets them? This will pass them on through my own code, in places where I don't want them..

> I start to question why you are disconnecting clients from outside of the server threads in the first place (ie, why are you running an external timer)? Are you not able to just use timeouts directly in the server's events instead?

A long time ago we discussed this. In MS-Windows there is no reliable way of getting a 'timeout' of an open TCP connection. I can't remember the details, but it just did not work properly.  Lets not go there again.
The 5-second 'ping-pong' system I now use between the Server and the Clients (both Windows and Android devices) makes sure that if something goes wrong, either the Server OR the Client will disconnect, and attempt a new re-connect. And this has worked perfectly in the many years. Including my Server code of disconnecting a Client when I lose connection. Except in this situation.

> > If a TCP event occurs (a TCP RESET for example), _where_ in the code can I add a debug log entry to report this?
> It won't be reported until the _next_ I/O operation on the socket sees the event and reports it back to Indy as a socket error.

That would be in the next 5 seconds, due to the Server sending a 'ping' to the Client. 
But in this case there is no exception raised to my code, as these are all logged.

> You should be able to attach files here.
There is no such option here, in the Issues of Github. I only see a 'Add a Link' option.

As said, the debug Log I posted are log entries I simply inserted into the actual Indy10 code. In 10 different Units by now. This is the Indy10 code. The only exception is my own 'DisconnectClient' procedure, which normally works perfectly.

I will show it here, but this is not where the problem is. The fact that I cannot get rid of the 'Context' has nothing to do with my code. At that point, it is already too late.

Note again this situation:
- In my code I get a Timeout after 7 'pings' send to the Client and not returned by the Client as a 'pong'.
- My 'DisconnectClient' is then called which (in the List) now calls "TIdContext(List[Index]).Binding.CloseSocket;"
- But nothing happens because it shows there is no connection: 
[23:37:08/948Z] DisconnectClient calls TIdContext(List[Index]).Binding.CloseSocket <<<<<<<
[23:37:08/948Z] IdSocketHandle.Disconnect
[23:37:08/948Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[23:37:08/948Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
[23:37:13/947Z] IdSocketHandle.CloseSocket: NOT HandleAllocated, Doing nothing. <<<<<<<<<<<<<<<

// In IdSocketHandle:
procedure TIdSocketHandle.CloseSocket;
begin
  FConnectionHandle.Enter;
  try
    if HandleAllocated then begin
    // Bart
      DLocalLog('IdSocketHandle.CloseSocket: HandleAllocated: Calling Disconnect',true);
      // Must be first, closing socket will trigger some errors, and they
      // may then call (in other threads) Connected, which in turn looks at
      // FHandleAllocated.
      FHandleAllocated := False;
      Disconnect;
      SetHandle(Id_INVALID_SOCKET);
    end else
    begin
      DLocalLog('IdSocketHandle.CloseSocket: NOT HandleAllocated, Doing nothing.',true);
    end;
  finally
    FConnectionHandle.Leave;
  end;
end;

**Comment by SARTrack** [comment]
Okay, new test update: It looks like this time the connection was still active, despite the fact that after 7 'pings' nothing was received from the Client:
D3: [17:43:52/095Z] [12345] KeepAliveTimerEvent: Port 2: Timeout for Client: Callsign=TESTPH, IP=192.168.50.110 PongCounter=7 Disconnecting on Index: 0
D3: [17:43:52/095Z] DisconnectClient: Disconnecting: Port=2  Callsign=TESTPH Index=0  Count=1
D3: [17:43:52/095Z] DisconnectClient calls TIdContext(List[Index]).Binding.CloseSocket
// Now in Indy:
[17:43:52/095Z] IdSocketHandle.CloseSocket: HandleAllocated: Calling Disconnect
[17:43:52/095Z] IdSocketHandle.Disconnect
[17:43:52/095Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
[17:43:52/095Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
// Here the Context List is first released, a short Sleep, then the List is checked again:
ERROR: [17:43:52/204Z] [12345] KeepAliveTimerEvent (Port2): DTBClient was not released! Index=0

So I have called CloseSocket, this then does call Disconnect but it is not doing anything else. 
Specifically, IdThread.Terminate is not called, or IdThread.Cleanup.  And IdIOHandlerSocket.Close is not called, and IdIOHandler.Close.... Why not?

**Comment by rlebeau** [comment]
> Also, are you telling me that I should 'raise' any and all Indy10 exeptions after my program gets them?

In the context of a `TIdTCPServer` event handler, you should typically not dismiss Indy's own exceptions.  Indy is designed to raise exceptions on disconnects and errors, and `TIdTCPServer` expects to receive exceptions to stop the socket thread when a client connection is closed or lost.

That being said, if you catch and dismiss Indy's exceptions, then you should either:

- raise your own exception instead and let it propagate into the server
- manually close the socket connection, which the server does check for.
- signal the calling thread directly to stop itself

The socket thread keeps looping, calling the server's `OnExecute` event on each iteration, until either:

- `TIdTCPConnection.Connected` returns False
- `TIdThread.Stopped` returns True
- An uncaught exception is raised

If you don't account for this, then `TIdTCPServer` doesn't know the client is done and the owning thread should stop.  Which could be why you are getting threads not terminating correctly, and thus eating up CPU cycles and not releasing their `Context`.

> A long time ago we discussed this. In MS-Windows there is no reliable way of getting a 'timeout' of an open TCP connection. I can't remember the details, but it just did not work properly. Lets not go there again. The 5-second 'ping-pong' system I now use between the Server and the Clients (both Windows and Android devices) makes sure that if something goes wrong, either the Server OR the Client will disconnect, and attempt a new re-connect. And this has worked perfectly in the many years. Including my Server code of disconnecting a Client when I lose connection.

Fair enough.

> That would be in the next 5 seconds, due to the Server sending a 'ping' to the Client. But in this case there is no exception raised to my code, as these are all logged.

Sending outgoing data on an abnormally lost connection can take a long time to report errors.  Until the OS times out internally and invalidates the connection, the OS happily continues to buffer outgoing data inside the socket.  Until the OS reports errors, Indy has no way to know the connection is lost.  Hence the need to use timeouts in your own code.

Even so, if you are sending the ping from outside of the server's events, then any error that would be raised would be in the context of the sending thread, not in the server's socket thread.  So, you may be responsible for having to delegate that fact into your event handler code so it can then shutdown properly.  Although, if you have a blocking read in progress, and a RESET occurs, the read should fail and raise an exception, which you would then have to act on if you catch it, otherwise let it propagate so the server can handle it.

> > You should be able to attach files here.
> There is no such option here, in the Issues of Github. I only see a 'Add a Link' option.

Hmm, I see an option to attach files.  You should, too:

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files

> As said, the debug Log I posted are log entries I simply inserted into the actual Indy10 code.

That really doesn't help me, though.  I can't see what you added and where you added it, and I can't see what operations you are calling into in the first place, or under what conditions.

> Note again this situation:
> 
>     * In my code I get a Timeout after 7 'pings' send to the Client and not returned by the Client as a 'pong'.
> 
>     * My 'DisconnectClient' is then called which (in the List) now calls "TIdContext(List[Index]).Binding.CloseSocket;"
> 
>     * But nothing happens because it shows there is no connection:
>       [23:37:08/948Z] DisconnectClient calls TIdContext(List[Index]).Binding.CloseSocket <<<<<<<
>       [23:37:08/948Z] IdSocketHandle.Disconnect
>       [23:37:08/948Z] IdSocketHandle.SetHandle calls FReadSocketList.Remove
>       [23:37:08/948Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
>       [23:37:13/947Z] IdSocketHandle.CloseSocket: NOT HandleAllocated, Doing nothing. <<<<<<<<<<<<<<<

This is normal. The socket handle is still open at this point.  So calling `TIdSocketHandle.CloseSocket()` will call `TIdSocketHandle.Disconnect()` to close the socket handle at the OS level, and then `TIdSocketHandle.SetHandle(Id_INVALID_SOCKET)` is called to clear Indy's references to that socket handle.  And since the new socket handle is `INVALID_SOCKET`, the `TIdSocketHandle` state is now disconnected and `TIdSocketHandle.HandleAllocated()` returns False.

`TIdTCPConnection.Connected()` (which `TIdTCPServer` looks at) returns False when `IOHandler.Connected()` is False.  And `TIdIOHandlerSocket.Connected()` returns False if `TIdSocketHandle.HandleAllocated()` is False *and the `InputBuffer` is empty*.  And `TIdIOHandlerStack.Connected()` performs a socket read to fill the `InputBuffer`, returning False if a disconnect occurs, otherwise it calls `TIdIOHandlerSocket.Connected()`.

**Comment by rlebeau** [comment]
> So I have called CloseSocket, this then does call Disconnect but it is not doing anything else. Specifically, IdThread.Terminate is not called, or IdThread.Cleanup. And IdIOHandlerSocket.Close is not called, and IdIOHandler.Close.... Why not?

At this point, I can't go off of your logging alone, it is just not sufficient to diagnose your issue.  In order to help you further, I need to see what your code is actually doing in the server's event handlers.  You can email me privately at remy@lebeausoftware.org, if you'd like.

**Comment by SARTrack** [comment]
Okay, I have put together a file which contains parts of my own code, and all the places in Indy where I added debug log entries.
These are all collected in a single TXT file.
I also added on the end the two situations after a 'DisconnectClient', the first one where it goes wrong the second one when it works normally.
[SARTrack-Code_Snippets.txt](https://github.com/IndySockets/Indy/files/11728612/SARTrack-Code_Snippets.txt)

**Comment by rlebeau** [comment]
You didn't include the code for `KeepAliveTimerEvent`

**Comment by SARTrack** [comment]
Here it is:
[KeepAliveTimerEvent.txt](https://github.com/IndySockets/Indy/files/11728796/KeepAliveTimerEvent.txt)

**Comment by SARTrack** [comment]
One more comment:
- It now also happens right here in my local WIFI, not via StarLink.
- It only seems to happen if I have two Android Clients connecting, every 2 or 5 minutes.
- When it does happen, (as per Log) often only one Client is connected at that time
- But having said that, this is after the 30 seconds timeout....
-

**Comment by SARTrack** [comment]
Two more things: 
1) I compared the debug log from the Server with that from the Android Client. 
And, according to the Android client Log, it requested a normal TCP disconnect after the synchonization with the Server was done. The Android client uses the IdTCPClient. So on that side I did not see anything abnormal. 
Only after it attempted to connect again 5 minutes later, the Server refused because it thought the client was still connected. I can see all this in both logs.
2) The special "DLocalLog()" system I use during these tests, are not based on PostMessage as is normally the case in my program, but they are saved to file on a line-by-line basis, using a "TCriticalSection" to prevent conflics from the various Threads.
As a consequence this has impact on the (timing) behaviour in Indy. 
And I have now noticed, that the problem is now occuring almost immediately, while previousely it sometimes would take hours for it to occur.
I think IMHO  the problem is Timing related. And it occurs when certain things are called in the 'wrong' order in Indy. 
Your opinion?

**Comment by rlebeau** [comment]
I've been out of town the past couple of days, so I haven't had a chance to look at your code yet. I'll try to get to that shortly.

**Comment by SARTrack** [comment]
I spend the last few days working solidly on this problem. All my other work is on hold.
My biggest problem is that I do not uderstand which parts of the code are calling other parts, which makes it nearly impossible to trace where it fails.
Please have a look at this last Log. 
[Indy-20230616A.txt](https://github.com/IndySockets/Indy/files/11763324/Indy-20230616A.txt)
Then look at this code:

procedure TIdThreadWithTask.Run;
begin
  if not FTask.DoRun then begin
    DLocalLog('IdThreadWithTask.Run calls Stop');
    Stop;
  end;
end;

This is the code which actually Stops the TiDThread. But it calls FTask.DoRun:

function TIdTask.DoRun: Boolean;
begin
  Result := Run;
end;
That is impossible, so there must be an Overwrite somewhere. But I cannot find it anywhere in the code.
This is NOT called in the case where it fails to terminate the TiDThread.

**Comment by SARTrack** [comment]
You say "your TIdTCPServer" but the problem is, it is in fact not my server, but the Indy10 server, written by you and many other people.  And I am spending massive amounts of time debugging it, originally when this same problem occured *all the time* under Linux.
I do appreciate your help. 
So please look at the details for me:

[16:50:22/752Z] IdSocketHandle.SetHandle sets Id_SOCK_UNKNOWN
// It stops here completely.
// The next attempt to "TIdContext(List[Index]).Binding.CloseSocket" shows "HandleAllocated" not allocated, and stops there.

// This block shows the normal procedure
[16:47:38/321Z] IdIOHandlerStack.CheckForDisconnect: ClosedGracefully

Please advise at which critical places in threads/loops  CheckForDisconnect is called. Because then I can try to find out WHY it is no longer being called, and locate the 'stuck' loop.
The TCP connect was in limbo. I then force a Disconnect.  Most of this is in fact accompliced.
The last part which it does is:

procedure TIdSocketListWindows.Remove(AHandle: TIdStackSocketHandle);
var
  i: Integer;
begin
  Lock;
  try
    {
    IMPORTANT!!!

    Sometimes, there may not be a member of the FDSET.  If you attempt to "remove"
    an item, the loop would execute once.
    }
    DLocalLog('IdSocketListWindows.Remove: FFDSet.fd_count='+FFDSet.fd_count.ToString);
    if FFDSet.fd_count > 0 then
    begin
      for i:= 0 to FFDSet.fd_count - 1 do
      begin
        if FFDSet.fd_array[i] = AHandle then
        begin
          Dec(FFDSet.fd_count);
          FFDSet.fd_array[i] := FFDSet.fd_array[FFDSet.fd_count];
          FFDSet.fd_array[FFDSet.fd_count] := 0; //extra purity
          DLocalLog('IdSocketListWindows.Remove: Found AHandle at Index '+i.ToString);
          Break;
        end;//if found
      end;
    end;
  finally
    Unlock;
  end;
end;

This is done.  FFDSet.fd_count is now in fact zero.
So, this should then cause another part of the program discover that the connection has been terminated. How else can it get to the next step?
And if it is all based on CheckForDisconnect, where is it normally called from?

**Comment by SARTrack** [comment]
Sorry, I just discovered that CheckForDisconnect *is* called all the time, but the HandleAllocated is false, so I could not Log it, as it happens all the time.

**Comment by SARTrack** [comment]
Another day of debugging.
What happens is that at some point, after a Client disconnects, without any exception of Log entry, the TCPServer.Execute is no longer executed. It just stops.
Please have a look at this file, which shows the Log and the exited sourcecode for procedure TIdThread.Execute;
Please advise if I am looking at the wrong place.
[Indy-20230618A.txt](https://github.com/IndySockets/Indy/files/11780236/Indy-20230618A.txt)

**Comment by SARTrack** [comment]
Okay, after two weeks of debugging, found the problem to be in my own code. 
With the help of Remy's explanation of how the Indy system actually works, discovered that a CriticalSection somewhere else in the program was not released under certain intermittant situations. 
This in turn caused the OnExecute to lock up so the TIdContext.OnRun never returned. 
This in turn caused the traffic to stop, and an external Timer then tried to recover and kill the connection, but this failed because due to the TIdContext.OnRun being locked up, all kinds of other parts of the Indy system also did not execute...

**Comment by emrahkucukali** [comment]
@SARTrack How did you resolve the issue? If Execute method entered the loop?

**Comment by SARTrack** [comment]
@emrahkucukali The problem was in my own code this time, where it locked up the OnExecute loop. However, I have made an ugly hack in the Indy code to break out of a loop here:
procedure TIdScheduler.TerminateAllYarns;
This loop locks up the entire system, causes a 100% CPU load (but not at this point, somewhere else).
I added a Counter here, if if the loop count exceeds a certain number, it breaks out of the loop, allowing the TCPServer to terminate (and the entire VPS). Also decreased the wait() to 100ms.
Ugly hack, but I got in major problems because of this. 
Again, under normal circumstances it *should* not happen...

**Comment by rlebeau** [comment]
I wouldn't recommend that hack, since `TerminateAllYarns` is not meant to exit until all yarns are actually terminated. If you exit prematurely, you risk leaving the server in a bad state.  Also, the loop already sleeps on each iteration, so there shouldn't be high CPU usage while it waits (a `TEvent` or `TSemaphore` would be more efficient to wait on, but oh well).
